catanatron_core/catanatron/state.py

```diff
@@ -591,9 +599,90 @@ def apply_action(state: State, action: Action):
         # state.current_player_index stays the same
         state.current_prompt = ActionPrompt.PLAY_TURN
         state.playable_actions = generate_playable_actions(state)
+    elif action.action_type == ActionType.OFFER_TRADE:
+        state.is_resolving_trade = True
+        state.current_trade = (*action.value, state.current_turn_index)
+
+        # go in seating order; order won't matter because of "acceptees hook"
+        state.current_player_index = next(
+            i for i, c in enumerate(state.colors) if c != action.color
+        )  # cant ask yourself
+        state.current_prompt = ActionPrompt.DECIDE_TRADE
+
+        state.playable_actions = generate_playable_actions(state)
+    elif action.action_type == ActionType.ACCEPT_TRADE:
+        # add yourself to self.acceptees
+        index = state.colors.index(action.color)
+        new_acceptess = list(state.acceptees)
+        new_acceptess[index] = True  # type: ignore
+        state.acceptees = tuple(new_acceptess)
+
+        try:
+            # keep going around table w/o asking yourself or players that have answered
+            state.current_player_index = next(
+                i
+                for i, c in enumerate(state.colors)
+                if c != action.color and i > state.current_player_index
+            )
+            # .is_resolving_trade, .current_trade, .current_prompt, .acceptees stay the same
+        except StopIteration:
+            # by this action, there is at least 1 acceptee, so go to DECIDE_ACCEPTEES
+            # .is_resolving_trade, .current_trade, .acceptees stay the same
+            state.current_player_index = state.current_turn_index
+            state.current_prompt = ActionPrompt.DECIDE_ACCEPTEES
+
+        state.playable_actions = generate_playable_actions(state)
+    elif action.action_type == ActionType.REJECT_TRADE:
+        try:
+            # keep going around table w/o asking yourself or players that have answered
+            state.current_player_index = next(
+                i
+                for i, c in enumerate(state.colors)
+                if c != action.color and i > state.current_player_index
+            )
+            # .is_resolving_trade, .current_trade, .current_prompt, .acceptees stay the same
+        except StopIteration:
+            # if no acceptees at this point, go back to PLAY_TURN
+            if sum(state.acceptees) == 0:
+                reset_trading_state(state)
```

Comment:
this feels like you want an "else", something like:

```python
curr = state.current_player_index
for i, c in enumerate(state.colors, curr + 1):
   if c != action.color:
      state.current_player_index = i
      break
  else:
     if sum(state.acceptees) == 0:
          reset_trading_state(state)
````
  (haven't tested that)

That said, how do you handle looping around? That is, if `current_player_index = 4`, and `len(players) == 5`, wouldn't this not give the player any chance to trade?

Reply:
I understand we don't have to loop around since when this phase is triggered by a OFFER_TRADE action it starts `current_layer_index` at `0`. 
 See https://github.com/bcollazo/catanatron/blob/feature/trading/catanatron_core/catanatron/state.py#L606.

And so, we only have to make sure it goes through the end of the table. If `len(players) == 5`, `len(state.colors) == 5`, and so 4 should be yielded by `i for i, c in enumerate(state.colors)` at some point?